File: bucket/q1Bucket.py
# initArr = [275, 87, 426, 61, 409, 170, 677, 503, 5, 1234]
initArr = [275, 87, 426, 61, 409, 170, 677, 503]
max_digits = len(str(max(initArr)))

for digit_place in range(max_digits):
    buckets = [[] for _ in range(10)]
    for number in initArr:
        digit = (number // (10 ** digit_place)) % 10
        buckets[digit].append(number)
    print(buckets)
    
    initArr = [num for bucket in buckets for num in bucket]

print(initArr)

File: bucket/q1Bucket.java
package bucket;
import java.util.*;

public class q1Bucket {
    public static void main(String[] args) {
        int[] initArr = {275, 87, 426, 61, 409, 170, 677, 503};
        int max_digits = Integer.toString(Arrays.stream(initArr).max().getAsInt()).length();

        for (int digit_place = 0; digit_place < max_digits; digit_place++) {
            List<List<Integer>> buckets = new ArrayList<>();
            for (int i = 0; i < 10; i++) {
                buckets.add(new ArrayList<>());
            }
            for (int number : initArr) {
                int digit = (number / (int) Math.pow(10, digit_place)) % 10;
                buckets.get(digit).add(number);
            }
            System.out.println(buckets);
            int index = 0;
            for (List<Integer> bucket : buckets) {
                for (int num : bucket) {
                    initArr[index++] = num;
                }
            }
        }

        System.out.println(Arrays.toString(initArr));
    }
}
